Develop script for preserving Fedora datastream version history

Fedora maintains a copy of all previous versions of datastreams (aka, version history). Occasionally, @andrewjbtw must consult this history to troubleshoot problems. However, previous versions cannot be migrated to Cocina since they may contain bad/unmappable data.

A plan must be developed to provide access to that version history once Fedora has been decommissioned.

Possible options:
* Dump the version history to a file system in Fedora archive format.
* Use fedora_cache to write to a file system.

Testing of the archive-fedora utility in my dev environment against https://sul-dor-qa.stanford.edu/fedora worked pretty smoothly on ~900 items. Moving to testing on `sdr-deploy` against https://sul-dor-migrate.stanford.edu/fedora using a DRUID list generated from its MySQL demonstrated that the Fedora instance would work sometimes but then start emitting 500 errors for items that on previous runs worked fine. Eventually Fedora started throwing timeout errors.

Looking in the Fedora logs /opt/app/fedora/fedora-3.8.1/server/logs/fedora.log revealed many errors like:

```
WARN 2022-02-24 17:44:50.410 [ajp-bio-8009-exec-3] (WebApplicationExceptionMapper) javax.ws.rs.WebApplicationException: org.fcrepo.common.FaultException: System error opening input stream
    at org.fcrepo.server.rest.BaseRestResource.handleException(BaseRestResource.java:203)
    at org.fcrepo.server.rest.BaseRestResource.handleException(BaseRestResource.java:208)
    at org.fcrepo.server.rest.FedoraObjectsResource.getObjectExport(FedoraObjectsResource.java:272)
    at sun.reflect.GeneratedMethodAccessor44.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.apache.cxf.service.invoker.AbstractInvoker.performInvocation(AbstractInvoker.java:180)
    at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
    at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:192)
    at org.apache.cxf.jaxrs.JAXRSInvoker.invoke(JAXRSInvoker.java:100)
    at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:57)
    at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:93)
    at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:272)
    at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
    at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:239)
    at org.apache.cxf.transport.servlet.ServletController.invokeDestination(ServletController.java:248)
    at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:222)
    at org.apache.cxf.transport.servlet.ServletController.invoke(ServletController.java:153)
    at org.apache.cxf.transport.servlet.CXFNonSpringServlet.invoke(CXFNonSpringServlet.java:167)
    at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:286)
    at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doGet(AbstractHTTPServlet.java:211)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:624)
    at org.apache.cxf.transport.servlet.AbstractHTTPServlet.service(AbstractHTTPServlet.java:262)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:303)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:369)
    at org.fcrepo.security.http.AuthZHttpRequestAttributesFilter.doFilter(AuthZHttpRequestAttributesFilter.java:112)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:381)
    at org.springframework.security.web.access.channel.ChannelProcessingFilter.doFilter(ChannelProcessingFilter.java:109)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:381)
    at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:168)
    at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:237)
    at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:167)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:218)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:110)
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:498)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:169)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
    at org.apache.catalina.valves.AccessLogValve.invoke(AccessLogValve.java:962)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:445)
    at org.apache.coyote.ajp.AjpProcessor.process(AjpProcessor.java:190)
    at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:637)
    at org.apache.tomcat.util.net.JIoEndpoint$SocketProcessor.run(JIoEndpoint.java:316)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    at java.lang.Thread.run(Thread.java:750)
Caused by: org.fcrepo.common.FaultException: System error opening input stream
    at org.fcrepo.server.storage.lowlevel.akubra.AkubraLowlevelStorage.openInputStream(AkubraLowlevelStorage.java:493)
    at org.fcrepo.server.storage.lowlevel.akubra.AkubraLowlevelStorage.retrieve(AkubraLowlevelStorage.java:419)
    at org.fcrepo.server.storage.lowlevel.akubra.AkubraLowlevelStorage.retrieveObject(AkubraLowlevelStorage.java:180)
    at org.fcrepo.server.storage.lowlevel.akubra.AkubraLowlevelStorageModule.retrieveObject(AkubraLowlevelStorageModule.java:83)
    at org.fcrepo.server.storage.DefaultDOManager.getReader(DefaultDOManager.java:733)
    at org.fcrepo.server.management.DefaultManagement.stream(DefaultManagement.java:361)
    at sun.reflect.GeneratedMethodAccessor45.invoke(Unknown Source)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:498)
    at org.fcrepo.server.messaging.NotificationInvocationHandler.invoke(NotificationInvocationHandler.java:71)
    at com.sun.proxy.$Proxy8.stream(Unknown Source)
    at org.fcrepo.server.management.ManagementModule.stream(ManagementModule.java:302)
    at org.fcrepo.server.rest.FedoraObjectsResource.getObjectExport(FedoraObjectsResource.java:265)
    ... 50 more
Caused by: java.io.FileNotFoundException: /opt/app/fedora/fedora-3.8.1/data/objectStore/9b/info%3Afedora%2Fdruid%3Abb001xx7991 (Too many open files)
    at java.io.FileInputStream.open0(Native Method)
    at java.io.FileInputStream.open(FileInputStream.java:195)
    at java.io.FileInputStream.<init>(FileInputStream.java:138)
    at org.akubraproject.fs.FSBlob.openInputStream(FSBlob.java:102)
    at org.akubraproject.impl.BlobWrapper.openInputStream(BlobWrapper.java:93)
    at org.fcrepo.server.storage.lowlevel.akubra.AkubraLowlevelStorage.openInputStream(AkubraLowlevelStorage.java:488)
```

I double checked that archive-fedora's use of Faraday isn't sending keep-alive that might be confusing Fedora3. 

@kamchan and @tallenaz took a look at the migrate.stanford.edu and concluded that the Fedora instance might be in an unstable state, and need to be reindexed. They started that process on Feb 25.

As it stands (single threaded, 4.3M items) the archive-fedora program would need to run continuously for almost a month (2 req/sec, 24 days) to completely archive our Fedora instance.

While looking for many open files errors from Fedora @jcoyne noticed [migration-utils](https://github.com/fcrepo-exts/migration-utils) that has been  developed to migrate Fedora3 to Fedora6's OCFL filesystem layout. It appears to bypass the API and to talk directly to the filesystem. @andrewjbtw suggested that this might suit his needs. Perhaps we should call this task done for now, and create a new one to experiment with migration-utils?

--- a/bin/archive-fedora
+++ b/bin/archive-fedora
@@ -1,0 +1,36 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require_relative '../config/environment'
+require_relative '../lib/fedora_archiver'
+
+options = { input: 'druids.txt', archive: 'archive' }
+
+parser = OptionParser.new do |option_parser|
+  option_parser.banner = 'Usage: bin/archive-fedora [options]'
+  option_parser.on('-iFILENAME', '--input FILENAME', String, 'File containing list of druids (instead of druids.txt).')
+  option_parser.on('-dDRUIDS', '--druids DRUIDS', Array, 'List of druids (instead of druids.txt).')
+  option_parser.on('-aDIRECTORY', '--archive DIRECTORY', String, 'Directory to write archived versions to (instead of archive).')
+  option_parser.on('-l', '--lenient', 'Log and ignore error messages from Fedora')
+  option_parser.on('-h', '--help', 'Displays help.') do
+    puts option_parser
+    exit
+  end
+end
+
+Rails.logger = ActiveSupport::Logger.new('archive-fedora.log')
+Rails.logger.formatter = proc do |severity, time, _, msg|
+  "#{time.strftime('%Y-%m-%d %H:%M:%S')} - #{severity} - #{msg}\n"
+end
+Rails.logger.datetime_format = '%Y-%m-%d %H:%M:%S'
+
+parser.parse!(into: options)
+druids = options[:druids] || File.read(options[:input]).split
+
+archiver = FedoraArchiver.new(
+  druids: druids,
+  archive_dir: options[:archive],
+  lenient: options[:lenient]
+)
+
+archiver.run

--- a/lib/fedora_archiver.rb
+++ b/lib/fedora_archiver.rb
@@ -1,0 +1,84 @@
+# frozen_string_literal: true
+
+require 'zlib'
+require 'ruby-progressbar'
+
+# Archive Fedora Objects as FOXML to a druid-tree
+class FedoraArchiver
+  include Rubydora::FedoraUrlHelpers
+
+  def initialize(druids:, archive_dir:, lenient:)
+    @druids = druids
+    @lenient = lenient
+    @archive_dir = Pathname.new archive_dir
+  end
+
+  def run
+    bar = ProgressBar.create(title: 'archiving', total: druids.length)
+    druids.map do |druid|
+      write_foxml(druid, archive_dir)
+      bar.increment
+    end
+  end
+
+  private
+
+  attr_reader :druids, :archive_dir, :lenient
+
+  def write_foxml(druid, archive_dir)
+    obj_file = archive_dir + "#{dtree(druid)}.xml.gz"
+    obj_dir = obj_file.dirname
+
+    if obj_file.file?
+      Rails.logger.info("#{druid} already found at #{obj_file}")
+      return
+    end
+
+    begin
+      content = fetch_foxml(druid)
+      if content
+        FileUtils.mkdir_p obj_dir
+        Zlib::GzipWriter.open(obj_file) do |gzip_file|
+          gzip_file.write(content)
+        end
+        Rails.logger.info("saved #{druid} as #{obj_file}")
+      end
+    rescue Faraday::TimeoutError
+      FileUtils.rm(obj_file) if obj_file.file?
+      msg = "timeout error when fetching #{druid}"
+      Rails.logger.error(msg)
+      raise msg unless lenient
+    rescue StandardError => e
+      FileUtils.rm(obj_file) if obj_file.file?
+      raise e
+    end
+  end
+
+  def fetch_foxml(druid)
+    resp = connection.get(export_object_url(druid, { context: 'archive' }))
+    if resp.status != 200
+      msg = "getting #{druid} returned #{resp.status}"
+      Rails.logger.error(msg)
+      return if lenient
+
+      raise msg
+    end
+    resp.body
+  end
+
+  def dtree(druid)
+    id = druid.delete_prefix('druid:')
+    "#{id[0..2]}/#{id[3..5]}/#{id}"
+  end
+
+  def connection
+    @connection ||= Faraday::Connection.new(
+      url: Settings.fedora_url,
+      ssl: {
+        client_cert: OpenSSL::X509::Certificate.new(File.read(Settings.ssl.cert_file)),
+        client_key: OpenSSL::PKey.read(File.read(Settings.ssl.key_file))
+      },
+      headers: { 'Accept' => 'text/xml' }
+    )
+  end
+end

--- a/spec/spec_helper.rb
+++ b/spec/spec_helper.rb
@@ -4,6 +4,7 @@ require 'simplecov'
 SimpleCov.start :rails do
   add_filter '/spec/'
   add_filter '/lib/data_error_notifier.rb'
+  add_filter '/lib/fedora_archiver.rb'
   add_filter '/lib/fedora_cache.rb'
   add_filter '/lib/fedora_loader.rb'
   add_filter '/lib/report.rb'

